### ISSUE TITLE
Cifar breaking

### ISSUE THREAD

**Issue by saonim** [tensorflow/tensorflow#1277]
I am running cifar multiGPU example with different batch sizes and number of GPUs. The code is breaking this way:

tensorflow.python.framework.errors.OutOfRangeError: RandomShuffleQueue '_1_tower_0/shuffle_batch/random_shuffle_queue' is closed and has insufficient elements (requested 128, current size 0)
     [[Node: tower_0/shuffle_batch = QueueDequeueMany[component_types=[DT_FLOAT, DT_INT32], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/cpu:0"](tower_0/shuffle_batch/random_shuffle_queue, tower_0/shuffle_batch/n/_775)]]
     [[Node: tower_1/shuffle_batch/n/_664 = _HostSend[T=DT_INT32, client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:1", send_device_incarnation=1, tensor_name="edge_170_tower_1/shuffle_batch/n", _device="/job:localhost/replica:0/task:0/gpu:1"](tower_1/shuffle_batch/n)]]
Caused by op u'tower_0/shuffle_batch', defined at:
  File "lib/python2.7/site-packages/tensorflow/models/image/cifar10/cifar10_multi-gpu_train.py", line 224, in <module>
    tf.app.run()
  File "/home/saoni.m/tensorflow/lib/python2.7/site-packages/tensorflow/python/platform/default/_app.py", line 30, in run
    sys.exit(main(sys.argv))
  File "lib/python2.7/site-packages/tensorflow/models/image/cifar10/cifar10_multi-gpu_train.py", line 222, in main
    train()
  File "lib/python2.7/site-packages/tensorflow/models/image/cifar10/cifar10_multi-gpu_train.py", line 150, in train
    loss = tower_loss(scope)
  File "lib/python2.7/site-packages/tensorflow/models/image/cifar10/cifar10_multi-gpu_train.py", line 65, in tower_loss
    images, labels = cifar10.distorted_inputs()
  File "/home/saoni.m/tensorflow/lib/python2.7/site-packages/tensorflow/models/image/cifar10/cifar10.py", line 119, in distorted_inputs
    batch_size=FLAGS.batch_size)
  File "/home/saoni.m/tensorflow/lib/python2.7/site-packages/tensorflow/models/image/cifar10/cifar10_input.py", line 153, in distorted_inputs
    min_queue_examples, batch_size)
  File "/home/saoni.m/tensorflow/lib/python2.7/site-packages/tensorflow/models/image/cifar10/cifar10_input.py", line 104, in _generate_image_and_label_batch
    min_after_dequeue=min_queue_examples)
  File "/home/saoni.m/tensorflow/lib/python2.7/site-packages/tensorflow/python/training/input.py", line 496, in shuffle_batch
    return queue.dequeue_many(batch_size, name=name)
  File "/home/saoni.m/tensorflow/lib/python2.7/site-packages/tensorflow/python/ops/data_flow_ops.py", line 287, in dequeue_many
    self._queue_ref, n, self._dtypes, name=name)
  File "/home/saoni.m/tensorflow/lib/python2.7/site-packages/tensorflow/python/ops/gen_data_flow_ops.py", line 319, in _queue_dequeue_many
    timeout_ms=timeout_ms, name=name)
  File "/home/saoni.m/tensorflow/lib/python2.7/site-packages/tensorflow/python/ops/op_def_library.py", line 664, in apply_op
    op_def=op_def)
  File "/home/saoni.m/tensorflow/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1834, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/home/saoni.m/tensorflow/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1043, in __init__
    self._traceback = _extract_stack()

The values that are passed to shuffle_batch():

```
 images, label_batch = tf.train.shuffle_batch(
      [image, label],
      batch_size=batch_size,
      num_threads=num_preprocess_threads,
      capacity=min_queue_examples + 3 * batch_size,
      min_after_dequeue=min_queue_examples)
```

are:
batch size 128, num_threads 16, capacity 20384, min_after_deque 20000

**Comment by girving** [comment]
@josh11b: Any ideas here?

**Comment by josh11b** [comment]
Does the CIFAR example set an epoch limit?  Looks like this might be
poor handling of running to the end in the CIFAR code?  That is the
normal end-of-input exception.

**Comment by girving** [comment]
It has `FLAGS.max_steps`, so looks like yes.

**Comment by josh11b** [comment]
Maybe @shlens would know more about the Cifar example? Is there a standard main loop now that handles this error?

**Comment by shlens** [comment]
If this is an issue due to poor handling of running to the end, then yes there is a better way to handle this. 
Namely, one would need to update [cifar10_multi_gpu_train.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/image/cifar10/cifar10_multi_gpu_train.py) to employ the [Supervisor](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/supervisor.py).

@sherrym would be the expert for upgrading this.

In the mean time, I would regard this error as a "non-error" as the training session merely did not shut down properly.

**Comment by hanskrupakar** [comment]
Hey, I'm using Ubuntu 16.04, CUDA 7.5, CuDNN v4, Tensorflow 0.8.0 with a single NVIDIA Geforce 650M 2GB GPU to train. 

I am implementing the same network with cifar100 as input instead of the cifar10 that is specified. I had the same exact error at the same exact line. If you traced the flow back to the source of the error, I think it lies in the cifar10_input.py in line number 149. 

Inside the distorted_inputs function in cifar10_input.py, I have changed this line to input cifar100 binary instead of cifar10.

    filenames = [os.path.join(data_dir, 'train.bin')] 

instead of

    filenames = [os.path.join(data_dir, 'data_batch_%d.bin' % i)
for i in xrange(1, 6)]

The error happened for me because there was some problem with the loop back mechanism of the string_input_producer function. It doesn't go back to the start of the file upon reaching EOF of the last file in the list filenames.

So, I changed the code to:
  filenames = [os.path.join(data_dir, 'train.bin') for _ in xrange(10000)]

It works now! Try doing the same for your code.
